### DISCUSSION TITLE
relationship between C/C++ Extension and cython

### DISCUSSION THREAD

**Discussion by ardeal** [pypa/setuptools#4028] [Q&A] [answered]
Hi,

Please check the following 2 pieces of code. 
in the first code, we don't use cython, but in the second, we use cython.
my questions are:
what is the difference of the following code?
`ext_modules=[]` and `ext_modules=cython()` both works, does this mean that cython return a list which includes Extension?
is there any doc about the explanation of the upper question?

```Python
setup(
	name = 'hello'
    ext_modules=[
        Extension('foo',
                  glob(path.join(here, 'src', '*.c')),
                  libraries = [ 'rt' ],
                  include_dirs=[numpy.get_include()])
    ]
)
```

```Python
setup(
	name = 'hello'
    ext_modules=cython([
        Extension('foo',
                  glob(path.join(here, 'src', '*.c')),
                  libraries = [ 'rt' ],
                  include_dirs=[numpy.get_include()])
    ])
)

```

**Comment by abravalheri** [comment]
Hi @ardeal, as far as I am aware there is no `cython` function exposed by setuptools.
Also I don't understand why you are using a `cython` function with an extension written in C.

It is not clear for me how do those things play out... I would recommend having a look on the Cython docs: https://cython.readthedocs.io/en/latest/

**Reply by ardeal** [reply]
Hi @abravalheri ,

Sorry, I input wrong words in my post. in my post, `cython` should be `cythonize`. 
I do see some code with `cythonize` which works. 

my doubt is:  For some code there is `cythonize`, but for some other code, there isn't `cythonize`.

**Reply by abravalheri** [reply, marked as answer]
Setuptools natively supports .c files in Extensions. That is a built-in capability.
You don't use `cythonize` with C extensions.

You just use cythonize when you have `.pyx` files[^1], because it converts `.pyx` files into `.c` (have a look on the Cython docs for more details).

[^1]: kind of... if you use the `Extension` class properly, you might not need to use `cythonize` at all because setuptools will try to call itself cythonize on the `.pyx` files. I recommend giving it a try and see if that works, if not you can revert to cythonize.

**Reply by ardeal** [reply]
got it, thanks, @abravalheri 
The comments in your link help:
`you might not need to use cythonize at all because setuptools will try to call itself cythonize on the .pyx files`
